File: weather-station/python/tests/station_builder.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone

from weather_station import AlertThresholds, Station

from .fixed_clock import FixedClock


@dataclass
class _SeededReading:
    time: datetime
    temperature: float
    humidity: float
    wind_speed: float


class StationBuilder:
    def __init__(self) -> None:
        self._start_time = datetime(2026, 6, 15, 12, 0, 0, tzinfo=timezone.utc)
        self._thresholds: AlertThresholds | None = None
        self._seeded: list[_SeededReading] = []

    def starting_at(self, time: datetime) -> StationBuilder:
        self._start_time = time
        return self

    def with_thresholds(self, thresholds: AlertThresholds) -> StationBuilder:
        self._thresholds = thresholds
        return self

    def with_reading(
        self,
        temperature: float,
        humidity: float,
        wind_speed: float,
        at: datetime | None = None,
    ) -> StationBuilder:
        self._seeded.append(
            _SeededReading(at or self._start_time, temperature, humidity, wind_speed)
        )
        return self

    def build(self) -> tuple[Station, FixedClock]:
        clock = FixedClock(self._start_time)
        station = Station(clock, self._thresholds)
        for seed in self._seeded:
            clock.advance_to(seed.time)
            station.record(seed.temperature, seed.humidity, seed.wind_speed)
        return station, clock

File: weather-station/typescript/tests/StationBuilder.ts
import { Station } from '../src/Station.js';
import type { AlertThresholds } from '../src/AlertThresholds.js';
import { FixedClock } from './FixedClock.js';

interface SeededReading {
  time: Date;
  temperature: number;
  humidity: number;
  windSpeed: number;
}

export class StationBuilder {
  private _startTime: Date = new Date(Date.UTC(2026, 5, 15, 12, 0, 0));
  private _thresholds: AlertThresholds = {};
  private _seededReadings: SeededReading[] = [];

  startingAt(time: Date): this {
    this._startTime = time;
    return this;
  }

  withThresholds(thresholds: AlertThresholds): this {
    this._thresholds = thresholds;
    return this;
  }

  withReading(temperature: number, humidity: number, windSpeed: number, at?: Date): this {
    this._seededReadings.push({
      time: at ?? this._startTime,
      temperature,
      humidity,
      windSpeed,
    });
    return this;
  }

  build(): { station: Station; clock: FixedClock } {
    const clock = new FixedClock(this._startTime);
    const station = new Station(clock, this._thresholds);
    for (const { time, temperature, humidity, windSpeed } of this._seededReadings) {
      clock.advanceTo(time);
      station.record(temperature, humidity, windSpeed);
    }
    return { station, clock };
  }
}
